syne_tune/optimizer/schedulers/searchers/dyhpo/hyperband_dyhpo.py

```diff
@@ -0,0 +1,234 @@
+# Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+from typing import List, Dict, Any, Tuple
+from numpy.random import RandomState
+import logging
+from collections import Counter
+
+from syne_tune.optimizer.schedulers.hyperband_promotion import (
+    PromotionRungSystem,
+)
+from syne_tune.optimizer.schedulers.searchers.dyhpo.dyhpo_searcher import (
+    DynamicHPOSearcher,
+    KEY_NEW_CONFIGURATION,
+)
+
+logger = logging.getLogger(__name__)
+
+
+DEFAULT_SH_PROBABILITY = 0.25
+
+
+class ScheduleDecision:
+    PROMOTE_SH = 0
+    PROMOTE_DYHPO = 1
+    START_DYHPO = 2
+
+
+class DyHPORungSystem(PromotionRungSystem):
+    """
+    Implements the logic which decides which paused trial to promote to the
+    next resource level, or alternatively which configuration to start as a
+    new trial, proposed in:
+
+        | Wistuba, M. and Kadra, A. and Grabocka, J.
+        | Dynamic and Efficient Gray-Box Hyperparameter Optimization for Deep Learning
+        | https://arxiv.org/abs/2202.09774
+
+    We do promotion-based scheduling, as in
+    :class:`~syne_tune.optimizer.schedulers.hyperband_promotion.PromotionRungSystem`.
+    In fact, we run the successive halving rule in :meth:`on_task_schedule` with
+    probability ``probability_sh``, and the DyHPO logic otherwise, or if the SH
+    rule does not promote a trial. This mechanism (not contained in the paper)
+    ensures that trials are promoted eventually, even if DyHPO only starts new
+    trials.
+
+    Since :class:`~syne_tune.optimizer.schedulers.HyperbandScheduler` was designed
+    for promotion decisions to be separate from decisions about new configs, the
+    overall workflow is a bit tricky:
+
+    * In :meth:`FIFOScheduler._suggest`, we first call
+      :code:`promote_trial_id, extra_kwargs = self._promote_trial()`. If
+      ``promote_trial_id != None``, this trial is promoted. Otherwise, we call
+      :code:`config = self.searcher.get_config(**extra_kwargs, trial_id=trial_id)`
+      and start a new trial with this config. In most cases, :meth:`_promote_trial`
+      makes a promotion decision without using the searcher.
+    * Here, we use the fact that information can be passed from
```

Comment:
I know this does not sound simple. But the structure of `HyperbandScheduler` is pragmatic, trying to keep scheduling decisions separate from search, and the latter is model-based if anything. This works fine and is simple for lots of variants we already have.

Just DyHPO is different, in that the decision on what to resume is folded with what to start new, it is even a common decision. One could of course change the whole design of `HyperbandScheduler` to accomodate that, but that would look complicated for all the other variants.